### PR TITLE
Add translation tables for states of some countries including Japan

### LINKED ISSUE #513: Billing Prefecture is not valid when using a Japanese Address.
When going to PayPal from the cart page and selecting a Japanese address in PayPal. We get redirected back to the "Confirm your PayPal order" page in WooCommerce with the following address:

```
151-0053
TOKYO-TO City Street
Test Japan
Japan
```

When we click the place order button we get the following error:

```
Billing Prefecture is not valid. Please enter one of the following: Hokkaido, Aomori, Iwate, Miyagi, Akita, Yamagata, Fukushima, Ibaraki, Tochigi, Gunma, Saitama, Chiba, Tokyo, Kanagawa, Niigata, Toyama, Ishikawa, Fukui, Yamanashi, Nagano, Gifu, Shizuoka, Aichi, Mie, Shiga, Kyoto, Osaka, Hyogo, Nara, Wakayama, Tottori, Shimane, Okayama, Hiroshima, Yamaguchi, Tokushima, Kagawa, Ehime, Kochi, Fukuoka, Saga, Nagasaki, Kumamoto, Oita, Miyazaki, Kagoshima, Okinawa
```

This is because the prefecture/state `TOKYO-TO` doesn't match with the WooCommerce state Tokyo.

PayPal uses the following values for prefecture/state:

```
HOKKAIDO = Hokkaido
AOMORI-KEN = Aomori
IWATE-KEN = Iwate
MIYAGI-KEN = Miyagi
AKITA-KEN = Akita
YAMAGATA-KEN = Yamagata
FUKUSHIMA-KEN = Fukushima
IBARAKI-KEN = Ibaraki
TOCHIGI-KEN = Tochigi
GUNMA-KEN = Gunma
SAITAMA-KEN = Saitama
CHIBA-KEN = Chiba
TOKYO-TO = Tokyo
KANAGAWA-KEN = Kanagawa
NIIGATA-KEN = Niigata
TOYAMA-KEN = Toyama
ISHIKAWA-KEN = Ishikawa
FUKUI-KEN = Fukui
YAMANASHI-KEN = Yamanashi
NAGANO-KEN = Nagano
GIFU-KEN = Gifu
SHIZUOKA-KEN = Shizuoka
AICHI-KEN = Aichi
MIE-KEN = Mie
SHIGA-KEN = Shiga
KYOTO-FU = Kyoto
OSAKA-FU = Osaka
HYOGO-KEN = Hyogo
NARA-KEN = Nara
WAKAYAMA-KEN = Wakayama
TOTTORI-KEN = Tottori
SHIMANE-KEN = Shimane
OKAYAMA-KEN = Okayama
HIROSHIMA-KEN = Hiroshima
YAMAGUCHI-KEN = Yamaguchi
TOKUSHIMA-KEN = Tokushima
KAGAWA-KEN = Kagawa
EHIME-KEN = Ehime
KOCHI-KEN = Kochi
FUKUOKA-KEN = Fukuoka
SAGA-KEN = Saga
NAGASAKI-KEN = Nagasaki
KUMAMOTO-KEN = Kumamoto
OITA-KEN = Oita
MIYAZAKI-KEN = Miyazaki
KAGOSHIMA-KEN = Kagoshima
OKINAWA-KEN = Okinawa
```

In WooCommerce we use the states as defined here: https://github.com/woocommerce/woocommerce/blob/master/i18n/states/JP.php

For other countries there is  a function to normalize the state to get around this mapping. See: https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/1.6.5/includes/class-wc-gateway-ppec-address.php#L345

To match the two up it looks like an array of something like this is needed (to convert between the PayPal value and WooCommerce):

```
'tokyo-to' => 'JP13'
```

### PR DESCRIPTION
**Pull request by menakas**
Fixes #513

As there is no translation table to normalize states of Japan, there was an error as described in the issue description. I added a translation table for Japan, using information from [here for WooCommerce](https://github.com/woocommerce/woocommerce/blob/master/i18n/states.php) and here for [PayPal](https://developer.paypal.com/docs/api/reference/state-codes/). In the process, I added translation tables for states of other countries listed in PayPal docs, but which do not have translation tables: Argentina, Indonesia, India, Mexico.

I faced some confusion while adding of translation tables for other countries.
1. India. There is an extra state 'Army Post Office' on PayPal side, which does not have an equivalent in WC. I have not included it in the translation table.

2. China. The number of states in WC(32) and PayPal(37) are not matching and some of them are quite similar(Shanxi vs. Shaanxi) that I have not included Chinese states in this PR.

3. Mexico. The code and names of 'DF' is confusing. `'DF' => __( 'Ciudad de M&eacute;xico', 'woocommerce' ),` is the entry in WC where as PayPal has a `Ciudad de México | CDMX` and a `Distrito Federal | DF`. At this point, I've commented out these two, but it appears they should both be combined into one. Wikipedia entry says this:

> On 29 January 2016, it ceased to be the Federal District (Spanish: Distrito Federal or D.F.) and is now officially known as Ciudad de México (or CDMX), with a greater degree of autonomy.

4. There is no state code for Thailand on PayPal side.

Any inputs on how to resolve these confusions are appreciated.

**Steps to Test**

1. Add a simple product to the cart.
2. View cart. https://d.pr/i/NUJilM
3. Click on PayPal smart button.
4. Log in into a buyer account.
5. Click on 'Change' against  the 'Ship to' address. Click on 'Add a new address 'https://d.pr/i/xjNF3c
6. Choose any prefecture of Japan https://d.pr/i/hq0xf8
7. Click on Continue.
8. You will get back to your site with a Japanese address in the Shipping address.
9. Now, click on 'Place Order'. https://d.pr/i/PeKYsU
10. On master, you will get an error 'Billing Prefecture is not valid. Please enter one of the following: Hokkaido, ...'. https://d.pr/i/PyNr0I On this branch, the order is placed successfully.

Closes #513

### CODE REVIEW

**Comment by james-allan** [comment]
> India. There is an extra state 'Army Post Office' on PayPal side, which does not have an equivalent in WC. I have not included it in the translation table.

That makes sense. Given WC doesn't list that as a possible store state location, it's fine not to include it given it won't need to be mapped.

> China. The number of states in WC(32) and PayPal(37) are not matching and some of them are quite similar(Shanxi vs. Shaanxi) that I have not included Chinese states in this PR.

It looks like these two examples are [legitimately different states](https://user-images.githubusercontent.com/8490476/73228159-81c6bb80-41c1-11ea-8fe4-506d4d8e20af.gif). 

I've had a go at mapping the WC supported states to PayPal codes. This is what I found: 

This is the initial map. These are the ones I could map cleanly based on the names used by both WC and PayPal:

```
array (
    'CN-YN' => 'CN1',
    'CN-BJ' => 'CN2',
    'CN-TJ' => 'CN3',
    'CN-HE' => 'CN4',
    'CN-SX' => 'CN5',
    'CN-LN' => 'CN7',
    'CN-JL' => 'CN8',
    'CN-HL' => 'CN9',
    'CN-SH' => 'CN10',
    'CN-JS' => 'CN11',
    'CN-ZJ' => 'CN12',
    'CN-AH' => 'CN13',
    'CN-FJ' => 'CN14',
    'CN-JX' => 'CN15',
    'CN-SD' => 'CN16',
    'CN-HA' => 'CN17',
    'CN-HB' => 'CN18',
    'CN-HN' => 'CN19',
    'CN-GD' => 'CN20',
    'CN-GX' => 'CN21',
    'CN-HI' => 'CN22',
    'CN-CQ' => 'CN23',
    'CN-SC' => 'CN24',
    'CN-GZ' => 'CN25',
    'CN-SN' => 'CN26',
    'CN-GS' => 'CN27',
    'CN-QH' => 'CN28',
    'CN-NX' => 'CN29',
    'CN-MO' => 'CN30',
    'CN-XJ' => 'CN32',
)
``` 

This leaves the following WC supported states which still need to be mapped. 

```
[CN6]  => Inner Mongolia
[CN31] => Tibet
```

with the following PayPal codes remaining: 

```
[Hong Kong SAR (en) Xianggang Tebiexingzhengqu (zh) 香港特别行政区 (Xiānggǎng Tèbiéxíngzhèngqū) Special administrative region] => CN-H
[Nei Mongol Zizhiqu (mn) 内蒙古自治区 (Nèi Ménggǔ Zìzhìqū) Autonomous region] => CN-NM
[Taiwan Sheng 台湾省 (Táiwān Shěng) Province] => CN-TW
[Xizang Zizhiqu 西藏自治区 (Xīzàng Zìzhìqū) Autonomous region] => CN-XZ
```

Doing some googling of those remaining PP names, you can manually map them. 

Nei Mongol Zizhiqu => Inner Mongolia
Xizang Zizhiqu => Tibet

so those would be:

```
'CN-NM' => 'CN6',
'CN-XZ' => 'CN31',
```

If you then take those two out of the list of remaining PP ones you're only left with: `Hong Kong SAR (en)` and `Taiwan Sheng`. Both of those are listed separately by WC so that makes sense that they aren't mapped.

Screen Shot 2020-01-28 at 12 28 19 pm

Screen Shot 2020-01-28 at 12 29 27 pm

What do you think of this mapping?

> The number of states in WC(32) and PayPal(37) are not matching

I counted 34 at PayPal not 37. With 34 the only 2 missing in WC are the ones they list separately. 

> Mexico. The code and names of 'DF' is confusing. 'DF' => __( 'Ciudad de Mexico', 'woocommerce' ), is the entry in WC where as PayPal has a Ciudad de México | CDMX and a Distrito Federal | DF. At this point, I've commented out these two, but it appears they should both be combined into one. 

Yeah I think you're right about combining the two. My guess is WC hasn't updated their code since Mexico city changed to longer be DF. Given it's listed at Ciudad de Mexico in WC, we should map it to CDMX in PayPal. 

> There is no state code for Thailand on PayPal side.

I couldn't find anything on that. I think we have to just leave Thailand as it is.

**Comment by james-allan** [comment]
> I couldn't find anything on that. I think we have to just leave Thailand as it is.

I tried following the steps to test with a Thai address and I couldn't complete it. I just got a generic PayPal error. So I don't know if that's with us or PayPal 🤷‍♂ 

Screen Shot 2020-01-28 at 1 05 07 pm

**Comment by menakas** [comment]
Thanks @james-allan for bringing more clarity.

>I counted 34 at PayPal not 37. 
Ah yes, PayPal has only 34 states for China. 

>you're only left with: Hong Kong SAR (en) and Taiwan Sheng. Both of those are listed separately by WC 
Both are regions under dispute. Whether they are considered as Chinese states or separate countries in their own right is the dispute. So, WC and PayPal are taking opposite views.

>What do you think of this mapping?
Great work navigating all those states and mapping them.

I've made the small change to the Mexican states issue and added the translation table for the Chinese states.

Let me know what you think.

**Comment by menakas** [comment]
While closing #501 I thought I shall use the state names from the error and build the translation table for Spain. I did that and found that it works for state names without accents. 

When a state name has an accent, the `strtolower()` function is not able to convert them to lowercase. So, I decided to leave them as upper case(for example, 'a coruÑa'). As there are no codes specified for the Spanish states and these names are substitutes for the state codes, I decided not to find an alternative for converting these characters to lowercase.

**Comment by menakas** [comment]
I've added a last commit for Thai states as well. But, as you mentioned, it doesn't work for whatever reason. The error happens at the paypal window itself. So, if at all, it is not related to this PR.

**Comment by menakas** [comment]
Considering the `strtolower()` function with accented characters problem faced with Spanish states, I decided to check the same for the states of Argentina which I added earlier. I faced the same issue as Thailand.

**Comment by james-allan** [comment]
> I've added a last commit for Thai states as well. But, as you mentioned, it doesn't work for whatever reason. The error happens at the paypal window itself. So, if at all, it is not related to this PR.

So how did you find those codes?

**Comment by menakas** [comment]
I tried to use the names as codes. The other way is to print them from the code [here](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/master/includes/class-wc-gateway-ppec-address.php/#L638).

**Comment by james-allan** [comment]
> I tried to use the names as codes. 

Oh right. I forgot the key and value pairs were PayPal code => WC code. So `TH-37` is the WC code. I thought the `TH-` were the PayPal ones and you found them somewhere. 🤦‍♂ 

> I tried to use the names as codes. 

Yeah, I discovered that PayPal doesn't always use the codes. For example, entering an address in China just returns the state code `[PAYMENTREQUEST_0_SHIPTOSTATE] => FUJIAN` not `CN-FJ` what they list as the state code.

File: includes/class-wc-gateway-ppec-address.php [hunk 1]
```diff
@@ -581,61 +835,169 @@ public function normalizeState() {
...
+				'virgin islands'                          => 'VI',
+				// The next few derivations don't show up on the list, but seem obvious
+				'us virgin islands'                       => 'VI',
+				'u.s. virgin islands'                     => 'VI',
+				'u s virgin islands'                      => 'VI',
+				'u. s. virgin islands'                    => 'VI'
 			);
 		}
 
+		error_log( strtolower(trim($this->_state)));
```
Comment (james-allan) [inline includes/class-wc-gateway-ppec-address.php]: @menakas there's an error log that slipped in here.

Comment (menakas) [reply includes/class-wc-gateway-ppec-address.php]: Oops. Thanks James. I removed it.

**Comment by james-allan** [comment]
This LGTM. I confirmed the mappings for all except for Thailand as I couldn't get the codes and I trust that you mapped them correctly. 

💯🏅 @menakas

**Comment by abelbascu** [comment]
> While closing #501 I thought I shall use the state names from the error and build the translation table for Spain. I did that and found that it works for state names without accents.
> 
> When a state name has an accent, the `strtolower()` function is not able to convert them to lowercase. So, I decided to leave them as upper case(for example, 'a coruÑa'). As there are no codes specified for the Spanish states and these names are substitutes for the state codes, I decided not to find an alternative for converting these characters to lowercase.

@menakas I'm having this issue purchasing via Google Pay. The 'Billing Province is not valid'. My test state is 'Lérida'. I do the purchase on the cart page, I did remove the Checkout page as it's only digital downloads. How to solve this? I guess woocommerce gets Google Play billing address as 'LÉrida' and it compares it with the associated bank account billing address 'Lérida' and it sees a mismatch? Or it does look for the state in the mandatory province field in the checkout page even that the purchase funnel doesn't go through it? (not an expert coder, just found this after many days looking for an answer). I know this is not the place to ask for this, feel free to remove my comment. Please let me know where I can discuss this issue (i can provide a screenshot of the error in chrome's console if needed). Thanks.

EDIT 09JAN21:
I think i found the culprit. The saved address in my Google Play states 'Lérida' as the province, but the 'Billing Province not valid' error thrown by woocommerce displays all the provinces in Spain, among them 'Lleida' (Catalonia's language spelling for 'Lérida'). Maybe woocommerce is trying to compare 'Lérida == Lleida'? Can I change some PHP code to tell woocoomerce to compare 'Lleida' or 'Lérida' depending on what each credit card provider offers in their billing address options? 

EDIT 10JAN21: 
@menakas Please find my workaround here: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/880#issuecomment-757380420
This may help you: https://www.terrytsang.com/tutorial/woocommerce/add-modify-states-for-woocommerce-checkout-form/

**Pull request merged**